driver/topology_service.cc

```diff
@@ -186,6 +183,48 @@ std::shared_ptr<CLUSTER_TOPOLOGY_INFO> TOPOLOGY_SERVICE::get_topology(CONNECTION
     return cached_topology;
 }
 
+std::shared_ptr<CLUSTER_TOPOLOGY_INFO> TOPOLOGY_SERVICE::get_filtered_topology(std::shared_ptr<CLUSTER_TOPOLOGY_INFO> topology) {
+    if (!this->allowed_and_blocked_hosts) {
+        return topology;
+    }
+
+    std::set<std::string> allowed_list = this->allowed_and_blocked_hosts->get_allowed_host_ids();
+    std::set<std::string> blocked_list = this->allowed_and_blocked_hosts->get_blocked_host_ids();
+
+    const std::shared_ptr<CLUSTER_TOPOLOGY_INFO> filtered_topology = std::make_shared<CLUSTER_TOPOLOGY_INFO>();
+    if (allowed_list.size() > 0) {
+        for (const auto& host : topology->get_instances()) {
+            if (allowed_list.find(host->get_host_id()) != allowed_list.end()) {
+                filtered_topology->add_host(host);
+            }
+        }
+    }
+
+    if (blocked_list.size() > 0) {
+        for (const auto& host : filtered_topology->get_instances()) {
+            // Remove blocked hosts from the filtered_topology.
+            if (blocked_list.find(host->get_host_id()) != blocked_list.end()) {
+                filtered_topology->remove_host(host);
+            }
+        }
+    }
```

Comment:
Given that calling `find` on an empty list will always return `end()`, I think this block can be simplified as follows which eliminates traversing `filtered_topology->get_instances()` twice.

```c++
for (const auto& host : topology->get_instances()) {
  if (allowed_list.find(host->get_host_id()) != allowed_list.end()) {
    filtered_topology->add_host(host);
  }
  if (blocked_list.find(host->get_host_id()) != blocked_list.end()) {
    filtered_topology->remove_host(host);
  }
}
```